ReSwift/CoreTypes/Middleware.swift

```diff
@@ -1,14 +1,99 @@
-    (DispatchFunction?, @escaping GetState) -> (@escaping DispatchFunction) -> DispatchFunction
+    internal let transform: (GetState, DispatchFunction, Action) -> [Action]
+
+    /// Create a blank slate Middleware.
+    public init() {
+        self.transform = { [$2] }
+    }
+
+    /**
+     Initialises the middleware with a transformative function.
+     
+     - parameter transform: The function that will be able to modify passed actions.
+     */
+    internal init(_ transform: @escaping (GetState, DispatchFunction, Action) -> [Action]) {
+        self.transform = transform
+    }
+
+    /**
+     Initialises the middleware by concatenating the transformative functions from
+     the middleware that was passed in.
+     */
+    public init(_ first: Middleware, _ rest: Middleware...) {
+        self = rest.reduce(first) {
+            $0.concat($1)
+        }
+    }
+
+    /// Safe encapsulation of side effects guaranteed 
+    /// not to affect the action being passed through the middleware.
+    public func sideEffect(_ effect: @escaping (GetState, DispatchFunction, Action) -> Void)
+        -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).map {
+                effect(getState, dispatch, $0)
+                return $0
+            }
+        }
+    }
+
+    /// Concatenates the transform function of the passed `Middleware` onto the callee's transform.
+    public func concat(_ other: Middleware) -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).flatMap {
+                other.transform(getState, dispatch, $0)
+            }
+        }
+    }
+
+    /// Concatenates the transform function onto the callee's transform.
+    public func map(_ transform: @escaping (GetState, Action) -> Action) -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).map {
+                transform(getState, $0)
+            }
+        }
+    }
+
+    /// Concatenates the transform function onto the callee's transform.
+    public func increase(_ transform: @escaping (GetState, Action) -> [Action]) -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).flatMap {
+                transform(getState, $0)
+            }
+        }
+    }
+
+    /// Concatenates the transform function onto the callee's transform.
+    public func flatMap(_ transform: @escaping (GetState, Action) -> Action?) -> Middleware {
+        return Middleware { getState, dispatch, action in
+            self.transform(getState, dispatch, action).flatMap {
+                transform(getState, $0)
+            }
+        }
+    }
+
+    /// Drop the Action if `predicate(action) != true`.
+    public func filter(_ predicate: @escaping (GetState, Action) -> Bool) -> Middleware {
```

Comment:
I think it would be a little clearer to use the argument name the Swift stdlib uses as well: `_ isIncluded:`. I think this is more self-explaining than `predicate`.